src/aws-cpp-sdk-core/source/client/RequestCompression.cpp

```diff
@@ -0,0 +1,229 @@
+        AWS_REQUEST_COMPRESSION_LOG_TAG,
+        "Compressed request requested in runtime without support for it");
+    return false;
+  }
+#else
+  // If there is no support to compress, always fail calls to this method.
+  AWS_LOGSTREAM_ERROR(
+      AWS_REQUEST_COMPRESSION_LOG_TAG,
+      "Compressed request requested in runtime without support for it");
+  AWS_UNREFERENCED_PARAM(input);     // silencing warning;
+  AWS_UNREFERENCED_PARAM(algorithm); // silencing warning;
+  return false;
+#endif
+}
+
+Aws::Utils::Outcome<std::shared_ptr<Aws::IOStream>, bool>
+Aws::Client::RequestCompression::uncompress(
+    std::shared_ptr<Aws::IOStream> input,
+    const CompressionAlgorithm &algorithm) const {
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+  if (algorithm == CompressionAlgorithm::GZIP) {
+    // calculating stream size
+    input->seekg(0, input->end);
+    auto streamSize = input->tellg();
+    input->seekg(0, input->beg);
+
+    AWS_LOGSTREAM_TRACE(AWS_REQUEST_COMPRESSION_LOG_TAG,
+                        "Uncompressing request of " << streamSize << " bytes.");
+
+    // Preparing output
+    std::shared_ptr<Aws::IOStream> output = Aws::MakeShared<Aws::StringStream>(
+        AWS_REQUEST_COMPRESSION_ALLOCATION_TAG);
+
+    // Prepare ZLIB to uncompress
+    int ret = Z_NULL;
+    z_stream strm;
+    std::unique_ptr<unsigned char[]> in(new unsigned char[ZLIB_CHUNK]);
+    std::unique_ptr<unsigned char[]> out(new unsigned char[ZLIB_CHUNK]);
+    // Cleaning Zlib stream state flags
+    strm.zalloc = Z_NULL;
+    strm.zfree = Z_NULL;
+    strm.opaque = Z_NULL;
+    strm.avail_in = 0;
+    strm.next_in = Z_NULL;
+
+    const int max_window_gzip = 31;
+    ret = inflateInit2(&strm, max_window_gzip);
+    if (ret != Z_OK) {
+      return false;
+    }
+
+    // Decompress
+    //  do {
+    // Fill the buffer
+    if (streamSize > ZLIB_CHUNK) {
+      // Skipping multi chunk for now
+      return false;
+    } else {
+      input->read(reinterpret_cast<char *>(in.get()), streamSize);
+
+      strm.avail_in = streamSize;
+      strm.next_in = in.get();
+      do {
+        // Run inflate on buffers
+        strm.avail_out = ZLIB_CHUNK;
+        strm.next_out = out.get();
+
+        ret = inflate(&strm, Z_NO_FLUSH);
+        // Catch errors
+        switch (ret) {
+        case Z_NEED_DICT:
+          AWS_LOGSTREAM_ERROR(
+              AWS_REQUEST_COMPRESSION_LOG_TAG,
+              "Compressed request failed to inflate with code: Z_NEED_DICT");
+          return false;
+        case Z_DATA_ERROR:
+          AWS_LOGSTREAM_ERROR(
+              AWS_REQUEST_COMPRESSION_LOG_TAG,
+              "Compressed request failed to inflate with code: Z_DATA_ERROR");
+          return false;
+        case Z_MEM_ERROR:
+          (void)inflateEnd(&strm);
+          AWS_LOGSTREAM_ERROR(
+              AWS_REQUEST_COMPRESSION_LOG_TAG,
+              "Compressed request failed to inflate with code: Z_MEM_ERROR");
+          return false;
+        }
```

Comment:
what if ret does not fall into any of the switch statement?

Reply:
Then it is not an error.